### DISCUSSION TITLE
How can i use a ref to my html dom element into the store?

### DISCUSSION THREAD

**Discussion by NirmalRaj-byjus** [pmndrs/zustand#1833] [Q&A]
In my store,'Remotestream' is a key value pair as below type,

type RemoteStream = {
	uid: string;
	isVideoEnabled: boolean;
	isAudioEnabled: boolean;
	name: string;
	streamRef: Ref<HTMLDivElement | null> | null;
	isVisible: boolean;
	isBlocked: boolean;
	isPinned: boolean;
	isFocused: boolean;
	isHandRaised: boolean;
	isNew: boolean;
};

In this object, streamRef is a prop contains ref to the div element, so during rendering i need this ref to be attached to the target div element. 

But I'm facing below error.
<img width="538" alt="Screenshot 2023-06-02 at 6 57 14 PM" src="https://github.com/pmndrs/zustand/assets/81288912/58597b7b-e406-490f-bad6-8a8d13ffa931">

How can i use this.. or whats the effective way of implementing it.

**Comment by dbritto-dev** [comment]
@NirmalRaj-byjus would you mind sharing us a basic example of your implementation on codesandbox?

**Reply by NirmalRaj-byjus** [reply]
I tried a similar approach here https://codesandbox.io/s/cool-lena-5vdyds?file=/src/App.js. But there is no error in codesandbox.

**Reply by NirmalRaj-byjus** [reply]
Adding sample here.

```
import { produce } from 'immer';
import { StateCreator } from 'zustand';
import { immer } from 'zustand/middleware/immer';

export type RemoteStreamsModel = {
	remoteStreams: { [key: string]: RemoteStream };
	addRemoteStreams: (streams: { [key: string]: RemoteStream }) => void;
	updateRemoteStream: (id: string, update: Partial<RemoteStream>) => void;
};

export const createRemoteStreamsSlice: StateCreator<RemoteStreamsModel, [], [['zustand/immer', never]], RemoteStreamsModel> = immer<RemoteStreamsModel>(
	(set) => ({
		remoteStreams: {},
		addRemoteStreams: (streams) =>
			set(
				produce((state) => {
					state.remoteStreams = { ...state.remoteStreams, ...streams };
				})
			),
		updateRemoteStream: (id: string, update: Partial<RemoteStream>) =>
			set(
				produce((state) => {
					state.remoteStreams[id] = { ...state.remoteStreams[id], ...update };
				})
			),
	})
);
```


Each remote stream is populated with the below value

```
{
		uid,
		isVideoEnabled: streamStatus?.isVideoEnabled || false,
		isAudioEnabled: streamStatus?.isAudioEnabled || false,
		name: details?.full_name || uid,
		profilePicURL: details?.profile_photo_url || '',
		backgroundColor: getStreamBackgroundColor(), //random color generator
		streamRef: createRef(),
		hasDetails: !!details,
		profilePhotoType: 'initial_avatar',
		index: 0,
		isVisible: false,
		isBlocked: false,
		isFocused: false,
		isPinned: false,
		isHandRaised: false,
		isNew: false,
		isBirthday: false,
		nameBackgroundColor: '',
		isLocal,
	};
```


And attaching this streamRef key to a div like below

```
	<div
	ref={streamRef} // if i comment this line, no error... else getting an error stated above
	key={uid}
> ... </div>
```

**Reply by dbritto-dev** [reply]
@NirmalRaj-byjus nice I'll take a look of that

**Reply by dbritto-dev** [reply]
@NirmalRaj-byjus here you go https://codesandbox.io/p/sandbox/compassionate-brattain-5hld7y?file=%2Fsrc%2FApp.tsx%3A15%2C17

BTW, you can't mutate the item directly you need to use a method to do that in my example I added a method to set up the ref on each todo using callback ref

**Reply by dannobytes** [reply]
Did you not run into this serialization issue when passing in a very large HTML element object into the store?

It's starting to feel like non-serializable objects are problematic after all, especially when you involve immer + redux devtools middleware 🤔 

![image](https://github.com/pmndrs/zustand/assets/751279/378b2b2a-c9a2-4321-b7ef-f339fa280dcb)

**Reply by dbritto-dev** [reply]
> Did you not run into this serialization issue when passing in a very large HTML element object into the store?
> 
> It's starting to feel like non-serializable objects are problematic after all, especially when you involve immer + redux devtools middleware 🤔 
> 
> ![image](https://github.com/pmndrs/zustand/assets/751279/378b2b2a-c9a2-4321-b7ef-f339fa280dcb)
> 
> 

@dannobytes yeah, that's the expected behavior that's why I wouldn't persist those ref you should always get the latest one on render. So I would exclude those using `partialize`.

**Reply by baishengmei** [reply]
Hello senior, I have encountered the same problem. However, the link provided above can no longer be opened. Could you please help me re-enter a link or post a picture? Thank you very much.

**Reply by dbritto-dev** [reply]
> Hello senior, I have encountered the same problem. However, the link provided above can no longer be opened. Could you please help me re-enter a link or post a picture? Thank you very much.

@thesuryavivek sure

**Reply by akoncic** [reply]
Hi @dbritto-dev , can you re-post your solution? The CodeSandbox link provided before no longer works. Thanks.

**Reply by C-o-d-e-C-o-w-b-o-y** [reply]
@akoncic here is what worked for me, using [ref callbacks](https://legacy.reactjs.org/docs/refs-and-the-dom.html#callback-refs):

In my zustand store:

```typescript
    settingsContainerRef: createRef() as MutableRefObject<HTMLDivElement>,
    setSettingsContainerEl: (element: HTMLDivElement | null) => {
      if (!element) {
        return;
      }
      const newRef = createRef() as MutableRefObject<HTMLDivElement>;
      newRef.current = element;
      set({ settingsContainerRef: newRef });
    },
 ```
 
 In my code:
 
 ```typescript
     <div
      ref={setSettingsContainerEl}
```

hope that helps 🙏

**Comment by C-o-d-e-C-o-w-b-o-y** [comment]
Here is what worked for me, using [ref callbacks](https://legacy.reactjs.org/docs/refs-and-the-dom.html#callback-refs):

In my zustand store:

```typescript
    settingsContainerRef: createRef() as MutableRefObject<HTMLDivElement>,
    setSettingsContainerEl: (element: HTMLDivElement | null) => {
      if (!element) {
        return;
      }
      const newRef = createRef() as MutableRefObject<HTMLDivElement>;
      newRef.current = element;
      set({ settingsContainerRef: newRef });
    },
 ```
 
 In my jsx:
 
 ```typescript
     <div
      ref={setSettingsContainerEl}
```

hope that helps 🙏

**Reply by dbritto-dev** [reply]
@NirmalRaj-byjus yep, that's the right way using ref callbacks

**Reply by dbritto-dev** [reply]
just be careful to not to persist that data because is huge you should always get the latest ref from the DOM

**Reply by krPlatypus** [reply]
To avoid the error "Converting circular structure to JSON", the following tasks are required.

```
  settingsContainerRef: createRef() as MutableRefObject<HTMLDivElement>,
    setSettingsContainerEl: (element: HTMLDivElement | null) => {
      if (!element) {
        return;
      }
      const newRef = createRef() as MutableRefObject<HTMLDivElement>;
      newRef.current = element;
      set({ settingsContainerRef: newRef });
    },

...
partialize: (state) => {
        const {  settingsContainerRef, ...rest } = state;
        return rest;
      },

```
